### PR TITLE
Numerically stable Matern32

### PR DESCRIPTION
**Pull request by nrontsis**
tldr: A numerically stable implementation of Matern32. This issue aims to address #685. A minimal example is not provided as the issues in the current implementation are rare, affecting e.g. ill-conditioned datasets, and iterative frameworks such as GPflowOpt.

The suggested implementation avoids the two issues existing in the current implementation:
1.  NaNs in the kernel outputs (see #421)
2.  Inf in the kernel's gradient (see #490)

Regarding (1): this is because the current implementation of `scaled_square_dist` can give negative results. ~~A new implementation of `scaled_square_dist` is introduced, taken from [here](https://stackoverflow.com/questions/37009647/compute-pairwise-distance-in-a-batch-without-replicating-tensor-in-tensorflow/40053188#40053188).~~ We use `tf.abs` (only in the forward pass) to get around this.

Regarding (2) this is due to the chain rule applied to `tf.sqrt`, which at zero has an infinite derivative. I avoid this by feeding analytically the gradient of `(1 + tf.sqrt(d2))*tf.exp(-tf.sqrt(d2)`, which has always bounded derivatives.

Tests are included that compare against the checking implementation.

### CODE REVIEW

**Comment by nrontsis** [comment]
Just to mention that the same idea could be applied to the Matern 1/2 or 5/2, Exponential and (with a small adaptation) Cosine Kernels.

File: gpflow/kernels.py [hunk 1]
```diff
@@ -250,22 +250,19 @@ def euclid_dist(self, X, X2):  # pragma: no cover
...
-        dist = -2 * tf.matmul(X, X2, transpose_b=True)
-        dist += tf.reshape(Xs, (-1, 1)) + tf.reshape(X2s, (1, -1))
+            X2 = X
+        else:
+            X2 = X2 / self.lengthscales
+
+        dist = tf.reduce_sum(tf.squared_difference(
+            tf.expand_dims(X, 1),
+            tf.expand_dims(X2, 0)), 2)
+
```
Comment (hughsalimbeni) [inline gpflow/kernels.py]: In my experience this is very much slower than the code it replaces

Comment (nrontsis) [reply gpflow/kernels.py]: Is this a concern? I thought that other operations e.g. Cholesky decomposition is the bottleneck.

Comment (hughsalimbeni) [reply gpflow/kernels.py]: It can be, especially when the input dimension is large. I think the gradient override is compatible with the existing squared distance computation, though.

Comment (nrontsis) [reply gpflow/kernels.py]: Okay. I this case we can get around the negative distances by adding a tf.abs which we will ignore in the gradients via a gradient override.

**Comment by nrontsis** [comment]
Now that I think it more,  I believe that a simpler hack would be to replace the euclidean distance calculation to:
```
r2 = self.scaled_square_dist(X, X2)
return tf.sqrt(tf.maximum(r2, 1e-12))
```
It shouldn't make such a difference, since `scaled_square_dist` is not so precise (i.e. it can negative values).

**Comment by hughsalimbeni** [comment]
Seems like a good way to go. Do you have a minimal example which exhibits the numerical instability?

**Comment by nrontsis** [comment]
No, I've experienced these issues while doing Bayesian Optimization and I don't have something simple at hand.

File: gpflow/kernels.py [hunk 2]
```diff
@@ -424,13 +424,28 @@ class Matern32(Stationary):
...
     @params_as_tensors
     def K(self, X, X2=None, presliced=False):
         if not presliced:
             X, X2 = self._slice(X, X2)
-        r = self.scaled_euclid_dist(X, X2)
-        return self.variance * (1. + np.sqrt(3.) * r) * \
-               tf.exp(-np.sqrt(3.) * r)
+        d = self.square_dist(X, X2)
+
+        with self.graph.gradient_override_map(
```
Comment (awav) [inline gpflow/kernels.py]: A bit more about *gradient_override_map*
From docs: https://www.tensorflow.org/api_docs/python/tf/Graph

gradient_override_map(op_type_map)
**EXPERIMENTAL**: A context manager for overriding gradient functions.

Comment (awav) [reply gpflow/kernels.py]: Does it mean there is no guarantee that this will work or supported in future by TensorFlow?

Comment (nrontsis) [reply gpflow/kernels.py]: Yeah, I guess this might not be future proof.

**Comment by awav** [comment]
Fantastic! Now you can convert it to the test. Make it this way (w.r.t [contributing](https://github.com/GPflow/GPflow/blob/master/contributing.md#code-quality-requirements)):

```
import pytest
from gpflow.test_utils import session_tf
from gpflow.kernels import Matern12, Matern32, Matern52, Exponential

class Datum:
    lengthscales = ...
    # cetera...

@pytest.mark.parametrize('kernel', [Matern12, Matern32, Matern52, Exponential]):
def test_error(session_tf, kernel):
    k = kernel(..., lengthscales=Datum.lengthscale, variance=Datum.variance)
    ... 
```
Check this [test](https://github.com/GPflow/GPflow/blob/master/tests/test_expectations.py) as an example.

File: gpflow/kernels.py [hunk 3]
```diff
@@ -424,11 +424,17 @@ class Matern32(Stationary):
     The Matern 3/2 kernel
     """
 
+    # Used to bypass gradients of tf.sqrt that can lead to infs.
+    tf.RegisterGradient("Matern32Grad")(
```
Comment (awav) [inline gpflow/kernels.py]: You can remove this.

File: tests/test_kerns.py [hunk 4]
```diff
@@ -43,6 +43,60 @@ def test_1d(self):
             self.assertTrue(np.allclose(gram_matrix, reference_gram_matrix))
 
 
+class Matern32Old(gpflow.kernels.Stationary):
```
Comment (awav) [inline tests/test_kerns.py]: No need in old implementation, once you cover your case with new test.

File: tests/test_kerns.py [hunk 5]
```diff
@@ -43,6 +43,60 @@ def test_1d(self):
...
+    @gpflow.decors.params_as_tensors
+    def K(self, X, X2=None, presliced=False):
+        if not presliced:
+            X, X2 = self._slice(X, X2)
+        r = self.scaled_euclid_dist(X, X2)
+        return self.variance * (1. + np.sqrt(3.) * r) * \
+               tf.exp(-np.sqrt(3.) * r)
+
+
+class TestMatern32(GPflowTestCase):
```
Comment (awav) [inline tests/test_kerns.py]: Same here, you can remove it. And it is not advised to use GPflowTestCase. Use pytest framework.

**Comment by nrontsis** [comment]
Thanks, all done.

File: tests/test_scaled_euclid_dist.py [hunk 6]
```diff
@@ -1,34 +1,46 @@
...
-    kernels = {'Matern12': Matern12, 'Matern32': Matern32,
-               'Matern52': Matern52, 'Exponential': Exponential,
-               'Cosine': Cosine}
-
-    k = kernels[kernel](Datum.D)
-
+    
+    k = kernel(Datum.D)
     K = k.compute_K_symm(Datum.X)
     assert not np.isnan(K).any(), 'There are NaNs in the output of the ' + kernel + ' kernel.'
```
Comment (nrontsis) [inline tests/test_scaled_euclid_dist.py]: kernel -> str(kernel) to avoid a TypeError.

File: tests/test_scaled_euclid_dist.py [hunk 7]
```diff
@@ -1,34 +1,46 @@
...
 
-    X = tf.placeholder(float_type)
+    X = tf.placeholder(settings.float_type)
     dK = session_tf.run(tf.gradients(k.K(X, X), X)[0], feed_dict={X: Datum.X})
     assert not np.isnan(dK).any(), 'There are NaNs in the gradient of the ' + kernel + ' kernel.'
```
Comment (nrontsis) [inline tests/test_scaled_euclid_dist.py]: Same here

**Comment by nrontsis** [comment]
@awav thanks alot for cleaning up the code. Is there anything else you would like to do? (apart from pushing the minor change in the assert message)

**Comment by awav** [comment]
Once tests pass, I will merge it.

**Comment by nrontsis** [comment]
Thanks! Last thing: added asserts for Infs as well, and added a forgotten __name__ in the first assert message.

**Comment by nrontsis** [comment]
However, a side effect of clipping at non-zero is getting zero gradient at zero distances for non-differentiable kernels like the Exponential (check its 1d gradient [here](http://www.wolframalpha.com/input/?i=derivative+exp(abs(x)))).

A more concise treatment would be to clip at zero for Matern12, Exponential and Cosine kernels but not for the Matern32 and Matern52 which are differentiable. I don't know if we ever prefer NaNs over zeros though.
